File: articles/case-study-integrated-assessment-and-sustainability-pathways/python/integrated_assessment_sustainability_pathways_workflow.py
#!/usr/bin/env python3
"""
Case study: integrated assessment and sustainability pathways.

Dependency-light workflow demonstrating:

1. Sustainability pathway comparison
2. Energy demand and clean energy transition
3. Annual and cumulative emissions
4. Climate stress and damages
5. Adaptation capacity
6. Land, water, equity, and sustainability diagnostics

All data are synthetic.
"""

from __future__ import annotations

from dataclasses import dataclass, asdict
from pathlib import Path
import csv
from typing import Any


ARTICLE_ROOT = Path(__file__).resolve().parents[1]
DATA = ARTICLE_ROOT / "data"
TABLES = ARTICLE_ROOT / "outputs" / "tables"


@dataclass(frozen=True)
class Pathway:
    name: str
    demand_growth: float
    efficiency_gain: float
    clean_growth_early: float
    clean_growth_late: float
    adaptation_investment: float
    transition_cost_factor: float
    equity_support: float
    ecological_constraint: float
    description: str


def clamp(value: float) -> float:
    return max(0.0, min(1.0, value))


def read_pathways(path: Path) -> list[Pathway]:
    pathways: list[Pathway] = []

    with path.open("r", newline="", encoding="utf-8") as handle:
        for row in csv.DictReader(handle):
            pathways.append(
                Pathway(
                    name=row["pathway"],
                    demand_growth=float(row["demand_growth"]),
                    efficiency_gain=float(row["efficiency_gain"]),
                    clean_growth_early=float(row["clean_growth_early"]),
                    clean_growth_late=float(row["clean_growth_late"]),
                    adaptation_investment=float(row["adaptation_investment"]),
                    transition_cost_factor=float(row["transition_cost_factor"]),
                    equity_support=float(row["equity_support"]),
                    ecological_constraint=float(row["ecological_constraint"]),
                    description=row["description"],
                )
            )

    return pathways


def read_csv_dicts(path: Path) -> list[dict[str, str]]:
    with path.open("r", newline="", encoding="utf-8") as handle:
        return list(csv.DictReader(handle))


def write_csv(path: Path, rows: list[dict[str, Any]]) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    if not rows:
        raise ValueError(f"No rows to write: {path}")

    fieldnames: list[str] = []
    for row in rows:
        for key in row:
            if key not in fieldnames:
                fieldnames.append(key)

    with path.open("w", newline="", encoding="utf-8") as handle:
        writer = csv.DictWriter(handle, fieldnames=fieldnames, extrasaction="ignore")
        writer.writeheader()
        writer.writerows(rows)


def simulate_pathway(pathway: Pathway, years: int = 40) -> list[dict[str, Any]]:
    demand = 1.00
    clean_share = 0.22
    cumulative_emissions = 0.0
    adaptation_capacity = 0.28
    rows: list[dict[str, Any]] = []

    for year in range(years + 1):
        clean_growth = pathway.clean_growth_early if year < 15 else pathway.clean_growth_late
        emissions_intensity = 0.72 * (1.0 - clean_share)
        annual_emissions = demand * emissions_intensity
        cumulative_emissions += annual_emissions

        climate_stress = clamp(0.18 + 0.018 * cumulative_emissions)
        adaptation_capacity = clamp(adaptation_capacity + pathway.adaptation_investment - 0.010 * climate_stress)

        climate_damages = 0.42 * climate_stress**2 * (1.0 - adaptation_capacity)
        transition_cost = pathway.transition_cost_factor * clean_growth * 4.0

        land_pressure = clamp(0.22 + 0.18 * demand + 0.25 * clean_share - 0.18 * pathway.ecological_constraint)
        water_stress = clamp(0.25 + 0.16 * demand + 0.34 * climate_stress - 0.14 * adaptation_capacity)
        equity_score = clamp(0.42 + 0.36 * pathway.equity_support - 0.18 * transition_cost - 0.22 * climate_damages)

        sustainability_score = (
            0.24 * equity_score
            + 0.20 * clean_share
            + 0.16 * adaptation_capacity
            - 0.15 * annual_emissions
            - 0.10 * climate_damages
            - 0.08 * land_pressure
            - 0.07 * water_stress
        )

        rows.append(
            {
                "pathway": pathway.name,
                "year": year,
                "energy_demand": round(demand, 6),
                "clean_energy_share": round(clean_share, 6),
                "emissions_intensity": round(emissions_intensity, 6),
                "annual_emissions": round(annual_emissions, 6),
                "cumulative_emissions": round(cumulative_emissions, 6),
                "climate_stress": round(climate_stress, 6),
                "adaptation_capacity": round(adaptation_capacity, 6),
                "climate_damages": round(climate_damages, 6),
                "transition_cost": round(transition_cost, 6),
                "land_pressure": round(land_pressure, 6),
                "water_stress": round(water_stress, 6),
                "equity_score": round(equity_score, 6),
                "sustainability_score": round(sustainability_score, 6),
                "land_breach": land_pressure > 0.72,
                "water_breach": water_stress > 0.72,
                "equity_breach": equity_score < 0.45,
            }
        )

        demand = demand * (1.0 + pathway.demand_growth - pathway.efficiency_gain)
        clean_share = clamp(clean_share + clean_growth)

    return rows


def summarize(rows: list[dict[str, Any]]) -> dict[str, Any]:
    final = rows[-1]

    def mean(key: str) -> float:
        return sum(float(row[key]) for row in rows) / len(rows)

    constraint_breach_count = sum(
        1
        for row in rows
        if bool(row["land_breach"]) or bool(row["water_breach"]) or bool(row["equity_breach"])
    )

    return {
        "pathway": final["pathway"],
        "final_clean_energy_share": final["clean_energy_share"],
        "cumulative_emissions": final["cumulative_emissions"],
        "average_climate_damages": round(mean("climate_damages"), 6),
        "average_transition_cost": round(mean("transition_cost"), 6),
        "average_land_pressure": round(mean("land_pressure"), 6),
        "average_water_stress": round(mean("water_stress"), 6),
        "average_equity_score": round(mean("equity_score"), 6),
        "final_adaptation_capacity": final["adaptation_capacity"],
        "constraint_breach_count": constraint_breach_count,
        "average_sustainability_score": round(mean("sustainability_score"), 6),
    }


def main() -> None:
    pathways = read_pathways(DATA / "sustainability_pathways.csv")
    assumptions = read_csv_dicts(DATA / "model_assumptions.csv")
    diagnostics = read_csv_dicts(DATA / "diagnostic_definitions.csv")

    all_rows: list[dict[str, Any]] = []
    summary_rows: list[dict[str, Any]] = []

    for pathway in pathways:
        rows = simulate_pathway(pathway)
        all_rows.extend(rows)
        summary_rows.append(summarize(rows))

    summary_rows.sort(key=lambda row: float(row["average_sustainability_score"]), reverse=True)

    validation_rows = [
        {"check": "pathway_runs_created", "passed": len(all_rows) > 0, "value": len(all_rows)},
        {
            "check": "clean_share_normalized",
            "passed": all(0 <= float(row["clean_energy_share"]) <= 1 for row in all_rows),
            "value": "all_clean_shares_checked",
        },
        {
            "check": "adaptation_capacity_normalized",
            "passed": all(0 <= float(row["adaptation_capacity"]) <= 1 for row in all_rows),
            "value": "all_adaptation_values_checked",
        },
        {
            "check": "equity_score_normalized",
            "passed": all(0 <= float(row["equity_score"]) <= 1 for row in all_rows),
            "value": "all_equity_values_checked",
        },
        {
            "check": "land_pressure_normalized",
            "passed": all(0 <= float(row["land_pressure"]) <= 1 for row in all_rows),
            "value": "all_land_pressure_values_checked",
        },
        {
            "check": "water_stress_normalized",
            "passed": all(0 <= float(row["water_stress"]) <= 1 for row in all_rows),
            "value": "all_water_stress_values_checked",
        },
        {
            "check": "emissions_nonnegative",
            "passed": all(float(row["annual_emissions"]) >= 0 for row in all_rows),
            "value": "all_emissions_checked",
        },
        {"check": "summary_created", "passed": len(summary_rows) == len(pathways), "value": len(summary_rows)},
    ]

    write_csv(TABLES / "python_sustainability_pathways.csv", [asdict(pathway) for pathway in pathways])
    write_csv(TABLES / "python_integrated_assessment_timeseries.csv", all_rows)
    write_csv(TABLES / "python_integrated_assessment_summary.csv", summary_rows)
    write_csv(TABLES / "python_model_assumptions.csv", assumptions)
    write_csv(TABLES / "python_diagnostic_definitions.csv", diagnostics)
    write_csv(TABLES / "python_integrated_assessment_validation_checks.csv", validation_rows)

    print("Integrated assessment sustainability pathways workflow complete.")
    print(TABLES / "python_integrated_assessment_summary.csv")


if __name__ == "__main__":
    main()

File: articles/case-study-integrated-assessment-and-sustainability-pathways/r/integrated_assessment_sustainability_pathways_workflow.R
# integrated_assessment_sustainability_pathways_workflow.R
# Base R workflow: energy, emissions, climate stress, adaptation, land, water, equity, and sustainability pathways.

args <- commandArgs(trailingOnly = FALSE)
file_arg <- grep("^--file=", args, value = TRUE)

if (length(file_arg) > 0) {
  script_path <- normalizePath(sub("^--file=", "", file_arg[1]), mustWork = TRUE)
  article_root <- normalizePath(file.path(dirname(script_path), ".."), mustWork = TRUE)
} else {
  article_root <- normalizePath(getwd(), mustWork = TRUE)
}

data_dir <- file.path(article_root, "data")
tables_dir <- file.path(article_root, "outputs", "tables")
figures_dir <- file.path(article_root, "outputs", "figures")

dir.create(tables_dir, recursive = TRUE, showWarnings = FALSE)
dir.create(figures_dir, recursive = TRUE, showWarnings = FALSE)

pathways <- read.csv(file.path(data_dir, "sustainability_pathways.csv"), stringsAsFactors = FALSE)
assumptions <- read.csv(file.path(data_dir, "model_assumptions.csv"), stringsAsFactors = FALSE)
diagnostics <- read.csv(file.path(data_dir, "diagnostic_definitions.csv"), stringsAsFactors = FALSE)

clamp <- function(value) {
  max(0, min(1, value))
}

simulate_pathway <- function(row, years = 40) {
  demand <- 1.00
  clean_share <- 0.22
  cumulative_emissions <- 0
  adaptation_capacity <- 0.28
  rows <- data.frame()

  for (year in 0:years) {
    clean_growth <- ifelse(year < 15, row$clean_growth_early, row$clean_growth_late)
    emissions_intensity <- 0.72 * (1 - clean_share)
    annual_emissions <- demand * emissions_intensity
    cumulative_emissions <- cumulative_emissions + annual_emissions

    climate_stress <- clamp(0.18 + 0.018 * cumulative_emissions)
    adaptation_capacity <- clamp(adaptation_capacity + row$adaptation_investment - 0.010 * climate_stress)

    climate_damages <- 0.42 * climate_stress^2 * (1 - adaptation_capacity)
    transition_cost <- row$transition_cost_factor * clean_growth * 4.0

    land_pressure <- clamp(0.22 + 0.18 * demand + 0.25 * clean_share - 0.18 * row$ecological_constraint)
    water_stress <- clamp(0.25 + 0.16 * demand + 0.34 * climate_stress - 0.14 * adaptation_capacity)
    equity_score <- clamp(0.42 + 0.36 * row$equity_support - 0.18 * transition_cost - 0.22 * climate_damages)

    sustainability_score <-
      0.24 * equity_score +
      0.20 * clean_share +
      0.16 * adaptation_capacity -
      0.15 * annual_emissions -
      0.10 * climate_damages -
      0.08 * land_pressure -
      0.07 * water_stress

    rows <- rbind(
      rows,
      data.frame(
        pathway = row$pathway,
        year = year,
        energy_demand = demand,
        clean_energy_share = clean_share,
        emissions_intensity = emissions_intensity,
        annual_emissions = annual_emissions,
        cumulative_emissions = cumulative_emissions,
        climate_stress = climate_stress,
        adaptation_capacity = adaptation_capacity,
        climate_damages = climate_damages,
        transition_cost = transition_cost,
        land_pressure = land_pressure,
        water_stress = water_stress,
        equity_score = equity_score,
        sustainability_score = sustainability_score,
        land_breach = land_pressure > 0.72,
        water_breach = water_stress > 0.72,
        equity_breach = equity_score < 0.45,
        stringsAsFactors = FALSE
      )
    )

    demand <- demand * (1 + row$demand_growth - row$efficiency_gain)
    clean_share <- clamp(clean_share + clean_growth)
  }

  rows
}

all_runs <- data.frame()

for (i in seq_len(nrow(pathways))) {
  all_runs <- rbind(all_runs, simulate_pathway(pathways[i, ]))
}

summary_rows <- data.frame()

for (pathway_name in unique(all_runs$pathway)) {
  subset_rows <- all_runs[all_runs$pathway == pathway_name, ]

  summary_rows <- rbind(
    summary_rows,
    data.frame(
      pathway = pathway_name,
      final_clean_energy_share = subset_rows$clean_energy_share[nrow(subset_rows)],
      cumulative_emissions = subset_rows$cumulative_emissions[nrow(subset_rows)],
      average_climate_damages = mean(subset_rows$climate_damages),
      average_transition_cost = mean(subset_rows$transition_cost),
      average_land_pressure = mean(subset_rows$land_pressure),
      average_water_stress = mean(subset_rows$water_stress),
      average_equity_score = mean(subset_rows$equity_score),
      final_adaptation_capacity = subset_rows$adaptation_capacity[nrow(subset_rows)],
      constraint_breach_count = sum(subset_rows$land_breach | subset_rows$water_breach | subset_rows$equity_breach),
      average_sustainability_score = mean(subset_rows$sustainability_score),
      stringsAsFactors = FALSE
    )
  )
}

summary_rows <- summary_rows[order(-summary_rows$average_sustainability_score), ]

validation_checks <- data.frame(
  check = c(
    "pathway_runs_created",
    "clean_share_normalized",
    "adaptation_capacity_normalized",
    "equity_score_normalized",
    "land_pressure_normalized",
    "water_stress_normalized",
    "emissions_nonnegative",
    "summary_created"
  ),
  passed = c(
    nrow(all_runs) > 0,
    all(all_runs$clean_energy_share >= 0 & all_runs$clean_energy_share <= 1),
    all(all_runs$adaptation_capacity >= 0 & all_runs$adaptation_capacity <= 1),
    all(all_runs$equity_score >= 0 & all_runs$equity_score <= 1),
    all(all_runs$land_pressure >= 0 & all_runs$land_pressure <= 1),
    all(all_runs$water_stress >= 0 & all_runs$water_stress <= 1),
    all(all_runs$annual_emissions >= 0),
    nrow(summary_rows) == nrow(pathways)
  )
)

write.csv(pathways, file.path(tables_dir, "r_sustainability_pathways.csv"), row.names = FALSE)
write.csv(all_runs, file.path(tables_dir, "r_integrated_assessment_timeseries.csv"), row.names = FALSE)
write.csv(summary_rows, file.path(tables_dir, "r_integrated_assessment_summary.csv"), row.names = FALSE)
write.csv(assumptions, file.path(tables_dir, "r_model_assumptions.csv"), row.names = FALSE)
write.csv(diagnostics, file.path(tables_dir, "r_diagnostic_definitions.csv"), row.names = FALSE)
write.csv(validation_checks, file.path(tables_dir, "r_integrated_assessment_validation_checks.csv"), row.names = FALSE)

png(file.path(figures_dir, "r_integrated_assessment_emissions_pathways.png"), width = 1000, height = 700)
plot(
  NULL,
  xlim = range(all_runs$year),
  ylim = range(all_runs$cumulative_emissions),
  xlab = "Year",
  ylab = "Cumulative Emissions",
  main = "Integrated Assessment Sustainability Pathways"
)

for (pathway_name in unique(all_runs$pathway)) {
  subset_rows <- all_runs[all_runs$pathway == pathway_name, ]
  lines(subset_rows$year, subset_rows$cumulative_emissions, lwd = 2)
}

legend("topleft", legend = unique(all_runs$pathway), lwd = 2, cex = 0.70)
grid()
dev.off()

print(summary_rows)
print(validation_checks)
cat("R integrated assessment sustainability pathways workflow complete.\n")
